OCR text cleaning is a combination of working manually and working computationally.

### Load packages and input data

In [2]:
import pandas as pd
import numpy as np
import re
import docx
import os

In [4]:
# reference: https://stackoverflow.com/questions/25228106/how-to-extract-text-from-an-existing-docx-file-using-python-docx
# import docx as a list of lines
def getText(filename):
    doc = docx.Document(filename)
    fullText = []
    for para in doc.paragraphs:
        fullText.append(para.text)
    return fullText

- The auto converted docx files from OCR pdfs are not helpful. The main reason is that, there are two columns, header, and footers. 
- To get these data, I manually changed copied the data from the OCR pdfs and manually cleaned them into desired forms in docx.

In [167]:
index1 = getText("Boder_Index_1_clean.docx")
index2 = getText("Boder_Index_2_clean.docx")
index3 = getText("Boder_Index_3_clean.docx")

### Data Manipulation

In [168]:
# Delete all empty strings
index1 = list(filter(lambda a: a != "", index1))
index2 = list(filter(lambda a: a != "", index2))
index3 = list(filter(lambda a: a != "", index3))

In [103]:
### example
a = "xxx \n x"

In [104]:
a

'xxx \n x'

In [109]:
print("\n")

In [110]:
print("\\n")

\n


In [105]:
len(a) # \n is a single character

7

In [123]:
a[:4]

'xxx '

In [125]:
a[5:]

' x'

In [147]:
b = ["xxx\nyyy\nzzz", "iii", "jjj\nkkk"]

In [148]:
index_b=[]
i=0
while i <len(b):
    if "\n" in b[i]:
        print(i, "has \\n")
        j=0
        while j < len(b[i])-1: 
            print(j)
            if b[i][j]=="\n":
                print("b[i][j]")
                index_b.append(b[i][:j])
                b[i]=b[i][j+1:]
                print(b[i])
                j=-1
            j+=1
        i-=1
    else:
        index_b.append(b[i])
    i+=1

0 has \n
0
1
2
3
b[i][j]
yyy
zzz
0
1
2
3
b[i][j]
zzz
0
1
2 has \n
0
1
2
3
b[i][j]
kkk
0
1


In [149]:
index_b

['xxx', 'yyy', 'zzz', 'iii', 'jjj', 'kkk']

In [ ]:
### example ends

In [169]:
# Seperate \n into different elements
# return list of elements
# seperate by \n
# b is the list
def remove_line_char(b):
    index_b=[]
    i=0
    while i <len(b):
        if "\n" in b[i]:
            j=0
            while j < len(b[i])-1: 
                if b[i][j]=="\n":
                    index_b.append(b[i][:j])
                    b[i]=b[i][j+1:]
                    j=-1
                j+=1
            i-=1
        else:
            index_b.append(b[i])
        i+=1
    return index_b

In [170]:
index1 = remove_line_char(index1)
index2 = remove_line_char(index2)
index3 = remove_line_char(index3)

In [171]:
index1

['Accidents, (see also industrial accidents), 341, 2961, 3057, 3067',
 'Accusations, false, 1139, 2954',
 '“Actions”, see raids',
 'Actors, 1168 ',
 'Adolescents, 327 ',
 'Affidavits, for immigration, 131, 180',
 'Agronomists, 3035',
 'Agudah, 192, 203, 486, 1987, 2682, 2689, 2711, 2734 ',
 'Air Pollution, 46, 477, 2265, 2606, 2711 ',
 'Air raids, 173, 268, 304, 411, 476, 528, 565, 878, 958, 1545, 1611, 1638, 1665, 1733, 1863, 1925, 1940, 2166, 2278, 2377, 2567, 2595, 2659, 2731, 2959, 2996, 3087',
 'Anesthetics, (see also medical care), 307 ',
 'Anschluss, Austrian, 1138 ',
 'Anti-Semitism, 233, 283 ',
 'Appells, (roll calls in camps), 70, 88, 101, 126, 164, 262, 296, 343, 385, 399, 463, 544, 621, 846, 873, 1106, 1215, 1237, 1299, 1225, 1454, 1593, 1604, 1653, 1674, 1714, 1855, 1878, 1887, 1931, 1984, 2056, 2074, 2167, 2238, 2376, 2507, 2579, 2624, 2640, 2651 ',
 'Armies, (indexed by nationalities) (see also military operations) ',
 'Armies, allied, 177, 269 ',
 'Architects, 1137 Arti

In [172]:
index2

['Allerich, (?), 1107 ',
 'Alma - Ata, (formerly Vernyi), misspelled Almata, 214 ',
 'Altengrube, 1083',
 'Americans, 553, 857, 1609, 1728, 2985, 3069, 3099 ',
 'Ampfing, (?), 305 ',
 'Andrychow, 370',
 'Antolowka, (see Antonowka) ',
 'Antonowka, 3011 ',
 'Antwerp, 182, 456',
 'Arabs, 1900',
 'Argentina, 2589',
 'Ariege Department, France, 457 ',
 'Aschaffenburg, 454 ',
 'Aschersleben, 2657 ',
 'Atlanta, Georgia, 313 ',
 'Augustow, Poland, 320 ',
 'Auschwitz, (see also Buna), 17, 26, 46, 53, 64, 73, 104, 117, 142, 159, 261, 275, 287, 344, 358, 375, 389, 403, 458, 475, 510, 547, 585, 604, 796, 849, 1089, 1164, 1214, 1297, 1383, 1415, 1520, 1584, 1717, 1764, 1827, 2049, 2137, 2161, 2194, 2226, 2273, 2333, 2356, 2562, 2515, 2626, 2636, 2646, 2652, 2714, 2756, 2792, 2843, 2879, 2901, 2928, 3029, 3035, 3103 ',
 'Austria, 322, 459, 1137, 1501, 1983, 2455, 2564, 2960 ',
 'Austrians, 1614 ',
 'Baden, 451 ',
 'Baku, 2672 ',
 'Baranovichi, 369',
 'Barash, (?), 3074',
 'Barta, 81',
 'Barth, Siles

In [173]:
index3

['Altoff, Herr, 2540, 2561 ',
 'Anders, General, 2862 ',
 'Ansbacher, Rabbi, 453 ',
 'Augustine, Bishop Metropolitan, 670 ',
 'Bar, 215 ',
 'Baron, Mr.Baruch, 215 ',
 'Bassfreund, Juergen, (also Gast-freund), 276-318 ',
 'Battenschlager, Herr, 851 ',
 'Beck, Dr. Leo, 1167 ',
 'Bevin, Ernest, 2865 ',
 'Biebow, HanX, (Bubow ?), 1290, 1416, 1810 ',
 'Bielka, 469',
 'Binder, Janine, 1034-1063 ',
 'Birnbaum, Goldie, 2589 ',
 'Biro, Ludwig, 479 ',
 'Bisenhaus, Polia, 865-886 ',
 'Bloch Rabbi Meyer, 238 ',
 'Bloch Naomi, 239',
 'Bloch Pierre, 746',
 'Bloch Rabbi, (Telsh Yeshiva), 204 ',
 'Bock, (Guard), 2623 ',
 'Bodeman, Dr. SS, 2054 ',
 'Bramson Jacques, 732-826 ',
 'Bramson Leonty Moiseievich, 733 ',
 'Brandt, Herr, 975 ',
 'Braun Anna, 691-729 ',
 'Braun Julius, (see Kluever) ',
 'Brenner, (German commandant of Draney), 600 ',
 'Brigida, (an SS woman), 2713 ',
 'Bruger, Marianne, 340 ',
 'Brutzvik, Isaac, 215',
 'Bulow, Von, 472 ',
 'Bundy, Nelly, 1501-1580 ',
 'Caneman, Rabbi, 204 ',
 'C

- 'mistreatment, starvation, travel) should not begin with a new line
- '2349, 2389, 2538, 2615, 2882, 2966, 2911, 302a, 3099 ', is a line of number alone
- a, o are typos that need to be changed.

So, manually clean again. When repeat the processes, most of the mistakes should be cleaned.

In [207]:
# get rid of space after each string
for i in range(len(index1)):
    if index1[i][-1]==" ":
        index1[i]=index1[i][:-1]

In [209]:
for i in range(len(index2)):
    if index2[i][-1]==" ":
        index2[i]=index2[i][:-1]

In [210]:
for i in range(len(index3)):
    if index3[i][-1]==" ":
        index3[i]=index3[i][:-1]

#### Data Attributes:
- We want to divide them into three colums, namely "word", "note", and "num"; 
- Some of the entries do not have numbers of index;
- Some of the numbers are in range, i.e. XXXX-XXXX;
- Some of the rows end with space, others not;
- Some of the numbers are not actually numbers - OCR as characters - need further cleaning.

In [255]:
# make data frames
df1=pd.DataFrame(columns=['text', 'word', 'note', 'num'])
df2=pd.DataFrame(columns=['text', 'word', 'note', 'num'])
df3=pd.DataFrame(columns=['text', 'word', 'note', 'num'])
df1["text"]=index1
df2["text"]=index2
df3["text"]=index3

In [256]:
# change string into list
df1.text = df1.text.apply(lambda x: x.split(","))
df2.text = df2.text.apply(lambda x: x.split(","))
df3.text = df3.text.apply(lambda x: x.split(","))

In [257]:
# the first element is word
df1.word = df1.text.apply(lambda x: x[0])
df2.word = df2.text.apply(lambda x: x[0])
df3.word = df3.text.apply(lambda x: x[0])

In [258]:
# remove the first elements from the text column
df1.text = df1.text.apply(lambda x: x[1:])

In [259]:
df2.text = df2.text.apply(lambda x: x[1:])

In [260]:
df3.text = df3.text.apply(lambda x: x[1:])

In [248]:
# now, there is a space before everthing
df1.text[0]

[' (see also industrial accidents)', ' 341', ' 2961', ' 3057', ' 3067']

In [249]:
df2.text[55]

[' 67',
 ' 103',
 ' 125',
 ' 164',
 ' 299',
 ' 476',
 ' 1324',
 ' 1598',
 ' 1608',
 ' 2086',
 ' 2383',
 ' 2930']

In [250]:
df3.text[80]

[' Marsha', ' 352']

In [262]:
# remove the space before everything
for i in range(len(df1.text)):
    for j in range(len(df1.text[i])):
        df1.text[i][j] = df1.text[i][j][1:]

In [263]:
df1.text[0]

['(see also industrial accidents)', '341', '2961', '3057', '3067']

In [264]:
for i in range(len(df2.text)):
    for j in range(len(df2.text[i])):
        df2.text[i][j] = df2.text[i][j][1:]

In [265]:
for i in range(len(df3.text)):
    for j in range(len(df3.text[i])):
        df3.text[i][j] = df3.text[i][j][1:]

In [386]:
# change note and num columns to empty lists
df1.note = ''
df1.note = df1.note.apply(list)
df1.num = ''
df1.num = df1.num.apply(list)
df2.note = ''
df2.note = df2.note.apply(list)
df2.num = ''
df2.num = df2.num.apply(list)
df3.note = ''
df3.note = df3.note.apply(list)
df3.num = ''
df3.num = df3.num.apply(list)

In [ ]:
# try to change the list to type int
# if success, copy to column num
# if not success, copy to column note

In [387]:
for i in range(len(df1.text)):
    for j in range(len(df1.text[i])):
        try:
            k = int(df1.text[i][j])
            df1.num[i].append(k)
        except:
            l = df1.text[i][j]
            df1.note[i].append(l)

In [388]:
for i in range(len(df2.text)):
    for j in range(len(df2.text[i])):
        try:
            k = int(df2.text[i][j])
            df2.num[i].append(k)
        except:
            l = df2.text[i][j]
            df2.note[i].append(l)

In [389]:
for i in range(len(df3.text)):
    for j in range(len(df3.text[i])):
        try:
            k = int(df3.text[i][j])
            df3.num[i].append(k)
        except:
            l = df3.text[i][j]
            df3.note[i].append(l)

In [ ]:
### range of number should be move from note to num

In [307]:
# helper function: change range to list of numbers
f = "2456-9234"

In [308]:
list(range(2,5))

[2, 3, 4]

In [311]:
def to_range(str1):
    for i in range(len(str1)):
        if str1[i]=="-":
            a = int(str1[:i])
            b = int(str1[i+1:])
            return(list(range(a,b+1)))

In [312]:
to_range(f)

[2456,
 2457,
 2458,
 2459,
 2460,
 2461,
 2462,
 2463,
 2464,
 2465,
 2466,
 2467,
 2468,
 2469,
 2470,
 2471,
 2472,
 2473,
 2474,
 2475,
 2476,
 2477,
 2478,
 2479,
 2480,
 2481,
 2482,
 2483,
 2484,
 2485,
 2486,
 2487,
 2488,
 2489,
 2490,
 2491,
 2492,
 2493,
 2494,
 2495,
 2496,
 2497,
 2498,
 2499,
 2500,
 2501,
 2502,
 2503,
 2504,
 2505,
 2506,
 2507,
 2508,
 2509,
 2510,
 2511,
 2512,
 2513,
 2514,
 2515,
 2516,
 2517,
 2518,
 2519,
 2520,
 2521,
 2522,
 2523,
 2524,
 2525,
 2526,
 2527,
 2528,
 2529,
 2530,
 2531,
 2532,
 2533,
 2534,
 2535,
 2536,
 2537,
 2538,
 2539,
 2540,
 2541,
 2542,
 2543,
 2544,
 2545,
 2546,
 2547,
 2548,
 2549,
 2550,
 2551,
 2552,
 2553,
 2554,
 2555,
 2556,
 2557,
 2558,
 2559,
 2560,
 2561,
 2562,
 2563,
 2564,
 2565,
 2566,
 2567,
 2568,
 2569,
 2570,
 2571,
 2572,
 2573,
 2574,
 2575,
 2576,
 2577,
 2578,
 2579,
 2580,
 2581,
 2582,
 2583,
 2584,
 2585,
 2586,
 2587,
 2588,
 2589,
 2590,
 2591,
 2592,
 2593,
 2594,
 2595,
 2596,
 2597,
 2598,

In [313]:
[1,3,4] + [5,6,7]

[1, 3, 4, 5, 6, 7]

In [318]:
# example
e = ["xxx", "yyy", "245-6784", "zzz", "123=#49f", "123=-#49f"]

In [319]:
i = 0 
while i < len(e):
    if "-" in e[i]:
        j = 0
        while j < len(e[i])-1:
            if e[i][j]=="-":
                try:
                    int(e[i][j-1])
                    int(e[i][j+1])
                    print(e[i]) # df.num[x][i] = df.num[x][i] + to_range(e[i])
                    e.remove(e[i])
                    i-=1
                except:
                    0
            j+=1
    i+=1

245-6784


In [322]:
e

['xxx', 'yyy', 'zzz', '123=#49f', '123=-#49f']

In [ ]:
# end of example

In [330]:
df3.note[0][0][0]

'H'

In [331]:
len(df3.note[0])

1

In [390]:
for k in range(len(df1.note)):
    i = 0 
    while df1.note[k]!=None and i < len(df1.note[k]):
        if "-" in df1.note[k][i]:
            j = 0
            while df1.note[k]!=None and j < len(df1.note[k][i])-1:
                if df1.note[k][i][j]=="-":
                    try:
                        int(df1.note[k][i][j-1])
                        int(df1.note[k][i][j+1])
                        df1.num[k] = df1.num[k] + to_range(df1.note[k][i])
                        df1.note[k] = df1.note[k].remove(df1.note[k][i])
                        i-=1
                    except:
                        0
                j+=1
        i+=1

In [391]:
for k in range(len(df2.note)):
    i = 0 
    while df2.note[k]!=None and i < len(df2.note[k]):
        if "-" in df2.note[k][i]:
            j = 0
            while df2.note[k]!=None and j < len(df2.note[k][i])-1:
                if df2.note[k][i][j]=="-":
                    try:
                        int(df2.note[k][i][j-1])
                        int(df2.note[k][i][j+1])
                        df2.num[k] = df2.num[k] + to_range(df2.note[k][i])
                        df2.note[k] = df2.note[k].remove(df2.note[k][i])
                        i-=1
                    except:
                        0
                j+=1
        i+=1

In [392]:
for k in range(len(df3.note)):
    i = 0 
    while df3.note[k]!=None and i < len(df3.note[k]):
        if "-" in df3.note[k][i]:
            j = 0
            while df3.note[k]!=None and j < len(df3.note[k][i])-1:
                if df3.note[k][i][j]=="-":
                    try:
                        int(df3.note[k][i][j-1])
                        int(df3.note[k][i][j+1])
                        df3.num[k] = df3.num[k] + to_range(df3.note[k][i])
                        df3.note[k] = df3.note[k].remove(df3.note[k][i])
                        i-=1
                    except:
                        0
                j+=1
        i+=1

In [406]:
# change all the empty list and none into np.nan
for i in range(len(df1.note)):
    if df1.note[i]==None or df1.note[i]==[]:
        df1.note[i]=np.nan
    if df1.num[i]==None or df1.num[i]==[]:
        df1.num[i]=np.nan
for i in range(len(df2.note)):
    if df2.note[i]==None or df2.note[i]==[]:
        df2.note[i]=np.nan
    if df2.num[i]==None or df2.num[i]==[]:
        df2.num[i]=np.nan
for i in range(len(df3.note)):
    if df3.note[i]==None or df3.note[i]==[]:
        df3.note[i]=np.nan
    if df3.num[i]==None or df3.num[i]==[]:
        df3.num[i]=np.nan

In [407]:
df1

text  \
0    [(see also industrial accidents), 341, 2961, 3...   
1                                  [false, 1139, 2954]   
2                                          [see raids]   
3                                               [1168]   
4                                                [327]   
5                          [for immigration, 131, 180]   
6                                               [3035]   
7        [192, 203, 486, 1987, 2682, 2689, 2711, 2734]   
8                          [46, 477, 2265, 2606, 2711]   
9    [173, 268, 304, 411, 476, 528, 565, 878, 958, ...   
10                      [(see also medical care), 307]   
11                                    [Austrian, 1138]   
12                                          [233, 283]   
13   [(roll calls in camps), 70, 88, 101, 126, 164,...   
14   [(indexed by nationalities) (see also military...   
15                                  [allied, 177, 269]   
16                          [1137 Artists, 1168, 2962]   
17   [(see also passports), 210, 274, 527, 588, 833...   
18   [(in town squares, see also appells, raids), 4...   
19                       [(see also succor), 255, 587]   
20                                      [industry, 89]   
21   [(see also hygiene), 72, 95, 156, 169, 267, 29...   
22                   [Christian, (textile plant), 472]   
23                                 [explorers of, 909]   
24                                     [(see bribery)]   
25   [8, 107, 120, 551, 685, 721, 776, 812, 1089, 1...   
26                                              [1138]   
27                        [(see also communists), 642]   
28   [use and forbiddance of, 380, 1157, 1594, 1878...   
29   [(including death during) 47, 103, 12S, 300, 4...   
..                                                 ...   
288                                 [42, 66, 284, 330]   
289            [(see deportation, evacuation, travel)]   
290  [by foot, trains, etc.; death during, restrict...   
291  [(capos, shupos), 75, 538, 944, 1384, 1425, 23...   
292                                              [660]   
293  [(see also Maquis, partisans, resistance), 740...   
294  [including post-war shelters, 194, 195, 246, 2...   
295                            [828, 2950, 2965, 3041]   
296                [(see also rebellions), 2129, 2257]   
297            [alleged drinking of, 1109, 2270, 2608]   
298  [684, 715, 1107, 2221, 2412, 2455, 2520, 2567,...   
299                                             [1260]   
300                             [198, 236, 2026, 2863]   
301  [(see also lice), 551, 613, 705, 856, 918, 116...   
302     [(prayer in anticipation of death), 254, 1759]   
303      [post-war, 185, 2454, 2668, 2688, 2962, 3075]   
304  [(see also bathing), 88, 271, 303, 335, 355, 4...   
305  [drinking, (see also bathing, thirst, urine), ...   
306                        [322, 343, 464, 2257, 2433]   
307  [(see also appells, clothing, death, death mar...   
308          [548, 1083, 1213, 1562, 1875, 2541, 2955]   
309                                 [(bequests), 3097]   
310  [compulsory, conditions, excessive, pace, puni...   
311                                              [543]   
312                                              [851]   
313  [insignia, or patch, (see also tattoos), 27, 8...   
314                   [131, 199, 203, 237, 2683, 2696]   
315                      [220, 1133, 1756, 1797, 2633]   
316                       [(also Zionism), 2688, 3081]   
317            [(Jewish fighters’ organization), 2257]   

                          word  \
0                    Accidents   
1                  Accusations   
2                    “Actions”   
3                       Actors   
4                  Adolescents   
5                   Affidavits   
6                  Agronomists   
7                       Agudah   
8                Air Pollution   
9                    Air raids   
10                 Anesthetics   
11                   Anschluss   
12         

In [408]:
df2

text               word  \
0                                          [(?), 1107]           Allerich   
1          [(formerly Vernyi), misspelled Almata, 214]         Alma - Ata   
2                                               [1083]         Altengrube   
3             [553, 857, 1609, 1728, 2985, 3069, 3099]          Americans   
4                                           [(?), 305]            Ampfing   
5                                                [370]          Andrychow   
6                                    [(see Antonowka)]          Antolowka   
7                                               [3011]          Antonowka   
8                                           [182, 456]            Antwerp   
9                                               [1900]              Arabs   
10                                              [2589]          Argentina   
11                                       [France, 457]  Ariege Department   
12                                               [454]      Aschaffenburg   
13                                              [2657]       Aschersleben   
14                                      [Georgia, 313]            Atlanta   
15                                       [Poland, 320]           Augustow   
16   [(see also Buna), 17, 26, 46, 53, 64, 73, 104,...          Auschwitz   
17      [322, 459, 1137, 1501, 1983, 2455, 2564, 2960]            Austria   
18                                              [1614]          Austrians   
19                                               [451]              Baden   
20                                              [2672]               Baku   
21                                               [369]        Baranovichi   
22                                         [(?), 3074]             Barash   
23                                                [81]              Barta   
24                                 [Silesia, 478, 482]              Barth   
25                                              [1745]              Basel   
26                                        [2962, 3044]            Bavaria   
27                                               [855]           Bayreuth   
28       [(also Bendzin, Bendin), 130, 165, 155, 1744]             Bedzin   
29                                         [810, 3022]           Belgians   
..                                                 ...                ...   
398              [204, 219, 229, 518, 669, 2568, 2667]              Vilna   
399                                 [(?), lager, 2125]             Visnie   
400                                             [2550]      Vistula River   
401                                       [lager, 483]           Walldurn   
402  [12, 265, 833, 957, 1637, 1852, 1904, 2019, 22...             Warsaw   
403                                             [3027]        Warta River   
404                                             [3027]           Wartagau   
405                [lager, (possibly Waldenburg), 382]         Weidenburg   
406                [821, 1611, 1838, 1855, 1941, 2410]             Weimar   
407                                             [3029]         Westphalia   
408                                              [948]            Wetzlar   
409                                             [2765]          Wieliczka   
410                                         [245, 483]          Wiesbaden   
411                                       [lager, 372]            Wischau   
412                                              [368]         Wojslawice   
413                                          [(?), 49]            Wolbrom   
414                                       [lager, 475]          Wolfsberg   
415                                      [lager, 2389]         Wolkenheim   
416                                    [(see Breslau)]            Wroclaw   
417                                             [3023]             Wurzen   
418                                         [(

In [409]:
df3

text                        word  \
0                        [Herr, 2540, 2561]                      Altoff   
1                           [General, 2862]                      Anders   
2                              [Rabbi, 453]                   Ansbacher   
3                [Bishop Metropolitan, 670]                   Augustine   
4                                     [215]                         Bar   
5                          [Mr.Baruch, 215]                       Baron   
6    [Juergen, (also Gast-freund), 276-318]                  Bassfreund   
7                               [Herr, 851]              Battenschlager   
8                           [Dr. Leo, 1167]                        Beck   
9                            [Ernest, 2865]                       Bevin   
10      [HanX, (Bubow ?), 1290, 1416, 1810]                      Biebow   
11                                    [469]                      Bielka   
12                      [Janine, 1034-1063]                      Binder   
13                           [Goldie, 2589]                    Birnbaum   
14                            [Ludwig, 479]                        Biro   
15                         [Polia, 865-886]                   Bisenhaus   
16                                    [238]           Bloch Rabbi Meyer   
17                                    [239]                 Bloch Naomi   
18                                    [746]                Bloch Pierre   
19                   [(Telsh Yeshiva), 204]                 Bloch Rabbi   
20                          [(Guard), 2623]                        Bock   
21                           [Dr. SS, 2054]                     Bodeman   
22                                [732-826]             Bramson Jacques   
23                                    [733]  Bramson Leonty Moiseievich   
24                              [Herr, 975]                      Brandt   
25                                [691-729]                  Braun Anna   
26                          [(see Kluever)]                Braun Julius   
27     [(German commandant of Draney), 600]                     Brenner   
28                    [(an SS woman), 2713]                     Brigida   
29                          [Marianne, 340]                      Bruger   
..                                      ...                         ...   
253                         [Roma, 827-864]                Tcharnobroda   
254                  [John H., (Jack), 246]                    Thompson   
255                     [Helena, 2043-2111]                   Tischauer   
256                             [Udel, 162]                 Tshernetsky   
257                         [Totetch, 2045]                        Tuka   
258                     [Israel, 1784-1849]                   Unikowski   
259                          [General, 513]                      Vlasov   
260            [(capo in Buchenwald), 1857]                       Vogel   
261                    [Bernard, 1850-1902]                    Warsager   
262                                   [492]             Wasserman Rabbi   
263                                  [2650]          Wasserman Chaykele   
264                             [Herr, 473]                      Weigel   
265                     [Julian, 1279-1301]                    Weinberg   
266                          [Fischel, 478]                      Wiener   
267                                  [2549]                     Wilczyk   
268                      [Jacob, 1473-1499]                        Wilf   
269                [(U.S. Clergyman), 2983]                        Wolf   
270                             [2423-2497]                  Wolf Isaac   
271                                   [353]           Wolovitch Charles   
272                                   [353]              Wolovitch Jake   
273                            [Chaim, 252]                    Zagalski   
274                              [Mr., 228]                     

It is acceptable that some of the cells are empty, as some rows don't have notes and some other rows don't have numubers.

### Export Data and further cleaning

In [410]:
df1.drop(columns=['text']).to_csv("Boder_indexs_1_subjects_situations.csv", index=False)
df2.drop(columns=['text']).to_csv("Boder_indexs_2_geographical_ethic.csv", index=False)
df3.drop(columns=['text']).to_csv("Boder_indexs_3_persons_organizations.csv", index=False)

- There could be mistakes, as the original is ORC. After export the data, then manually check and clean.
- All the unsure data are marked in red and yellow in the docxs. Please refere to them for further data cleaning. Links:

https://drive.google.com/drive/folders/1MMMDtK53AaZk9Z1urs0yRqTYsQByCE_x

- The result cleaned csvs are here: 

https://docs.google.com/spreadsheets/d/1ZCsFaeh9_wG5Wro_lBBZ-oLBjDC_PtE0rQJdvF5-6iQ/edit#gid=1799247479

https://docs.google.com/spreadsheets/d/1_HYMfC8vZSuMTc-11xi6Q55BSj5zwA51J1JGQjMbJ2c/edit#gid=976540509

https://docs.google.com/spreadsheets/d/1y0MHI3lbydzB6fc8LJEY1edNIjnLDsjMosS63wIidyY/edit#gid=1527998046